src/types/redis_tdigest.cc

```diff
@@ -0,0 +1,416 @@
+  double TotalWeight() const { return static_cast<double>(meta_data_.total_weight); }
+  double Min() const { return meta_data_.minimum; }
+  double Max() const { return meta_data_.maximum; }
+  uint64_t Size() const { return meta_data_.merged_nodes; }
+
+ private:
+  const TDigestMetadata& meta_data_;
+  const std::vector<Centroid>& centroids_;
+};
+
+uint64_t constexpr kMaxElements = 1 * 1024;  // 1k doubles
+
+std::optional<rocksdb::Status> TDigest::Create(engine::Context& ctx, const Slice& digest_name,
+                                               const TDigestCreateOptions& options) {
+  auto ns_key = AppendNamespacePrefix(digest_name);
+  auto capacity = options.compression * 6 + 10;
+  capacity = ((capacity < kMaxElements) ? capacity : kMaxElements);
+  TDigestMetadata metadata(options.compression, capacity);
+
+  LockGuard guard(storage_->GetLockManager(), ns_key);
+  auto status = GetMetaData(ctx, ns_key, &metadata);
+  if (status.ok()) {
+    return {};
+  }
+
+  if (!status.IsNotFound()) {
+    return status;
+  }
+
+  auto batch = storage_->GetWriteBatchBase();
+  WriteBatchLogData log_data(kRedisTDigest);
+  if (auto status = batch->PutLogData(log_data.Encode()); !status.ok()) {
+    return status;
+  }
+
+  std::string metadata_bytes;
+  metadata.Encode(&metadata_bytes);
+  if (auto status = batch->Put(metadata_cf_handle_, ns_key, metadata_bytes); !status.ok()) {
+    return status;
+  }
+
+  return storage_->Write(ctx, storage_->DefaultWriteOptions(), batch->GetWriteBatch());
+}
+rocksdb::Status TDigest::Add(engine::Context& ctx, const Slice& digest_name, const std::vector<double>& inputs) {
+  auto ns_key = AppendNamespacePrefix(digest_name);
+  LockGuard guard(storage_->GetLockManager(), ns_key);
+
+  TDigestMetadata metadata;
+  if (auto status = GetMetaData(ctx, ns_key, &metadata); !status.ok()) {
+    return status;
+  }
+
+  auto batch = storage_->GetWriteBatchBase();
+  WriteBatchLogData log_data(kRedisTDigest);
+  if (auto status = batch->PutLogData(log_data.Encode()); !status.ok()) {
+    return status;
+  }
+
+  metadata.total_observations += inputs.size();
+  metadata.total_weight += inputs.size();
+
+  if (metadata.unmerged_nodes + inputs.size() <= metadata.capacity) {
+    if (auto status = appendBuffer(ctx, batch, ns_key, inputs, &metadata); !status.ok()) {
+      return status;
+    }
+    metadata.unmerged_nodes += inputs.size();
+  } else {
+    if (auto status = mergeCurrentBuffer(ctx, ns_key, batch, &metadata, &inputs); !status.ok()) {
+      return status;
+    }
+  }
+
+  std::string metadata_bytes;
+  metadata.Encode(&metadata_bytes);
+  if (auto status = batch->Put(metadata_cf_handle_, ns_key, metadata_bytes); !status.ok()) {
+    return status;
+  }
+
+  return storage_->Write(ctx, storage_->DefaultWriteOptions(), batch->GetWriteBatch());
+}
+
+rocksdb::Status TDigest::Quantile(engine::Context& ctx, const Slice& digest_name, const std::vector<double>& qs,
```

Comment:
What if "Buffer" not exists in this scenerio?

Reply:
Hi @mapleFU ,

It will try to merge buffer before dumping centroids.

So, if there is no buffer, the merging will be skipped and continue calculating quantile with centroids.

Best Regards,
Edward

Reply:
But still lock would be held and new log would be added?

Reply:
Hi @mapleFU ,

Thanks for your suggestion! 😊

I have added a check before try to merge buffer into centroids.
It will skip this action if unmerged nodes is empty.

Best Regards,
Edward